### DISCUSSION TITLE
Sunrise, sunset times are off by 4 hours (UTC versus local time)

### DISCUSSION THREAD

**Discussion by achalmers** [cosinekitty/astronomy#222] [Q&A] [answered]
First time using the library. Using riseset.c as a guide, Im getting sunrise too early and sunsets too late. Are there config settings I need to set up ? What test case would you suggest ? Im comparing output to https://www.timeanddate.com 

eg 

    //Test for Boston MA 
    astro_time = Astronomy_MakeTime(2022, 05, 19, 0, 0, 0) ; 
    astro_observer_t observer = Astronomy_MakeObserver(42.25, -70.0, 0.0);
    sunrise  = Astronomy_SearchRiseSet(BODY_SUN,  observer, DIRECTION_RISE, astro_time, 300.0);
    sunset   = Astronomy_SearchRiseSet(BODY_SUN,  observer, DIRECTION_SET,  astro_time, 300.0);
    moonrise = Astronomy_SearchRiseSet(BODY_MOON, observer, DIRECTION_RISE, astro_time, 300.0);
    moonset  = Astronomy_SearchRiseSet(BODY_MOON, observer, DIRECTION_SET,  astro_time, 300.0);ve

produces :
observer.longitude = -70.00
observer.latitude = 42.25
observer.height = 0.00
2022-05-19T00:00:00Z
Sunrise at : 2022-05-19T09:15:05Z
Sunset at : 2022-05-19T23:58:30Z
Moonrise at : 2022-05-19T03:26:54Z
Moonset at : 2022-05-19T12:08:03Z

whereas the website lists both correctly 4 hors earlier : sunrise at 05.19  and sunset at 8.03  

Appreciate pointers.
BTW this is being run on an ESP32 micro in the Arduino environment.

**Comment by cosinekitty** [comment, marked as answer]
Hi @achalmers and welcome to the project! Your code looks great, and everything is working correctly. The 4 hour discrepancy is because the times are reported in UTC, not local time. Astronomy Engine does not know anything about time zones. It does everything in UTC. It turns out time zones are extremely complicated: the boundary lines are complicated and the rules change frequently at the whim of governments around the world. I decided early on that time zone support is out of scope for Astronomy Engine.

Usually you can use your operating system (C/C++, Python), browser (JavaScript), or language runtime (C#) to convert to local time. For example, in the [linux_riseset.c](https://github.com/cosinekitty/astronomy/blob/master/demo/c/linux_riseset.c) demo, you can see an example of where I use the Linux function `localtime_r` to convert UTC to local time. I doubt this will help in your case, because it sounds like you are running on a system that doesn't even have an operating system. Chances are you will need to manually enter a time zone correction by subtracting 4 hours in the summer, 5 hours in the winter (daylight savings time on/off).

Also, if you can add a couple more decimal places of accuracy to your longitude and latitude, the predictions should match timeanddate.com within a minute or two.

That is great that you are running this in an Arduino environment! I am a big fan of Arduino and Raspberry Pi. How much RAM does your system have? I ask because I wondered whether the Astronomy Engine C code would fit in a microcontroller system.

**Reply by achalmers** [reply]
Ah, UT, I think that explains it. Yes no local time available. I updated lat and lon for more precision and got the below (~5mins of timeanddate.com :-) 
Sunrise at : 2022-05-20T09:13:56Z
Sunset at : 2022-05-20T23:59:47Z
Moonrise at : 2022-05-20T04:25:11Z
Moonset at : 2022-05-20T13:20:48Z

Im running on a TTGO with 2 ESP32 cores @240MHz , system has 4MB Flash 32MB RAM, this app is about 350k. Amazing device, so much functionality in a small inexpensive package.

I was inspired by (and give credit to ) the bodmer  tft libraries and inclusion of an orrery example that used the astronomy code that piqued my interest. Ive admired the meeus books and moshiers astronomy code for a long time . I'll try and upload a short video.

**Comment by achalmers** [comment]
https://user-images.githubusercontent.com/7535369/169564322-b57edfad-f768-457a-a181-62339f40049e.mp4

**Reply by cosinekitty** [reply]
That is really cool! Did they already have code for calculating planet positions? I'm curious what they were using to do that.

**Reply by achalmers** [reply]
I just started with the library so I dont know about 3d position (any pointerto docs/references ?), the orrery sample uses 
    ang = Astronomy_EclipticLongitude(body[i], astro_time);

**Reply by cosinekitty** [reply]
Oh, you are writing this from scratch then. Cool.  Your best bet is to use [Astronomy_HelioVector](https://github.com/cosinekitty/astronomy/tree/master/source/c#Astronomy_HelioVector) for each of the planets you are interested in. This gives you a 3D vector from the Sun to the planet. Call [Astronomy_GeoMoon](https://github.com/cosinekitty/astronomy/tree/master/source/c#astronomy_geomoontime--astro_vector_t) to get a vector from the center of the Earth to the center of the Moon. You can add Earth's vector to this Moon vector to get a heliocentric Moon vector.

The only thing is, these vectors are expressed in Earth equatorial coordinates. For a top-down view of the solar system, it might be better to convert these vectors to ecliptic coordinates (the Earth's orbital plane).  I will come back later (time for dinner, LOL) and explain how to do this conversion, if you are interested.

**Reply by achalmers** [reply]
Thx, sure Im interested.   I'll first play with the HelioVector and maybe see if autoscaling in each axis will show anything interesting on a small tft display. I doubt it.  I also do some programming with Unity for the Oculus Quest and Ive always wanted to do some kind of space simulation. Switching between the top down and 3D sounds interesting.

**Reply by cosinekitty** [reply]
I added a demo program [ecliptic_vector.c](https://github.com/cosinekitty/astronomy/blob/master/demo/c/ecliptic_vector.c) that shows how to calculate positions of the Sun, Moon, and planets in ecliptic coordinates. You could use the x and y coordinates and ignore the z coordinate, to plot this on your graphical display.

With some more work, you could also use the [Astronomy_Pivot](https://github.com/cosinekitty/astronomy/tree/master/source/c#Astronomy_Pivot) function to reorient the rotation matrix `rot` to show the solar system from any other angle you wanted. For example, using some kind of hardware control to tilt around the x-axis.

I hope this helps!

**Reply by achalmers** [reply]
Adding HelioVector was straightforward, seems to be working. Will add our moon and 3D graphics next.
Code:    
//  02 version uses Astronomy_HelioVector(body, time) ⇒ astro_vector_t  

  int i;
  int num_bodies = sizeof(body) / sizeof(body[0]);

   for (i = 0; i < num_bodies; ++i) {
    astro_vector_t AV = Astronomy_HelioVector(body[i], astro_time);   
    Serial.print("-----------"); Serial.print(i); Serial.print(" ");Serial.println(AV.status);
    Serial.print("x : ");Serial.println(AV.x);
    Serial.print("y : ");Serial.println(AV.y);
    Serial.print("z : ");Serial.println(AV.z);
  }
Produces :
2022-05-20T00:00:00Z
Sunrise at : 2022-05-20T09:13:56Z
Sunset at : 2022-05-20T23:59:47Z
Moonrise at : 2022-05-20T04:25:11Z
Moonset at : 2022-05-20T13:20:48Z
-----------0 0
x : 0.00
y : 0.00
z : 0.00
-----------1 0
x : -0.26
y : -0.34
z : -0.16
-----------2 0
x : 0.54
y : -0.43
z : -0.23
-----------3 0
x : -0.53
y : -0.79
z : -0.34
-----------4 0
x : 0.99
y : -0.87
z : -0.43
-----------5 0
x : 4.91
y : -0.63
z : -0.39
-----------6 0
x : 7.45
y : -5.89
z : -2.76
-----------7 0
x : 14.01
y : 12.76
z : 5.39
-----------8 0
x : 29.69
y : -3.11
z : -2.01
----------------------------
